### PR TITLE
Add support for session window

### PR DESCRIPTION
**Pull request by petrpan26**
# Session Windows Implementation for Quix Streams

This PR introduces **Session Windows** - dynamic windowing that groups events by activity sessions rather than fixed time intervals.

## Key Features
- **Activity-driven sessions**: Sessions extend automatically when events arrive within timeout period
- **Flexible configuration**: Configurable timeout and grace periods for late events
- **Full aggregation support**: Works with all existing aggregation functions (sum, count, mean, max, min, reduce, collect)
- **Both modes supported**: `.current()` for real-time updates, `.final()` for complete sessions
- **Closing strategies**: Both "key" and "partition" level session expiration

## Usage Examples
```python
# User session tracking (30min timeout, 5min grace)
sdf.session_window(timeout_ms=timedelta(minutes=30), grace_ms=timedelta(minutes=5)) \
   .agg(page_views=Count(), actions=Collect("action_type")) \
   .final()

# Real-time fraud detection
sdf.session_window(timeout_ms=timedelta(minutes=10)) \
   .agg(transaction_count=Count(), total_amount=Sum("amount")) \
   .current()
```

## Use Cases Enabled
- User activity tracking (web/app sessions)
- Fraud detection (transaction patterns)
- IoT device monitoring (activity sessions)
- Gaming analytics (player sessions)

## Testing & Documentation
- ✅ **32 comprehensive tests** covering all session window functionality
- ✅ **Complete documentation** added to `docs/windowing.md` with examples
- ✅ **Session merging scenarios** and edge cases tested
- ✅ **Multi-key session management** validated

## Technical Implementation
- Smart session extension logic with grace period support
- Efficient RocksDB-backed state management
- Proper transaction handling for concurrent access
- Memory-efficient session garbage collection

This enables sophisticated streaming analytics use cases that were previously difficult to implement with fixed-time windows.

### CODE REVIEW

File: docs/windowing.md [hunk 1]
```diff
@@ -9,7 +9,9 @@ With windows, you can calculate such aggregations as:
 - Total of website visitors for every hour
 - The average speed of a vehicle over the last 10 minutes
 - Maximum temperature of a sensor observed over 30 second ranges
-- Give an user a reward after 10 succesful actions 
+- Give an user a reward after 10 succesful actions
```
Comment (gwaramadze) [inline docs/windowing.md:12]:
```suggestion
- Give a user a reward after 10 successful actions
```

File: docs/windowing.md [hunk 2]
```diff
@@ -500,6 +502,220 @@ sdf = (
...
+
+```
+Time:    0    5    10   15   20   25   30   35   40   45   50
+Events:  A         B              C    D              E
+
+Timeout: 10 seconds
+Grace:   2 seconds
+
+Session 1: [0, 20] - Events A, B (B extends the session from A)
+Session 2: [25, 35] - Events C, D (D extends the session from C)
```
Comment (gwaramadze) [inline docs/windowing.md:528]: shouldn't that be `[25, 40]`?
```suggestion
Session 2: [25, 40] - Events C, D (D extends the session from C)  
```

File: quixstreams/dataframe/windows/definitions.py [hunk 3]
```diff
@@ -526,4 +530,74 @@ def _get_name(self, func_name: Optional[str]) -> str:
...
+
+        super().__init__(name, dataframe, on_late)
+
+        self._timeout_ms = timeout_ms
+        self._grace_ms = grace_ms
+
+    @property
+    def timeout_ms(self) -> int:
+        return self._timeout_ms
+
```
Comment (gwaramadze) [inline quixstreams/dataframe/windows/definitions.py:570]: I don't think this property is used anywhere
```suggestion
```

File: quixstreams/dataframe/windows/time_based.py [hunk 4]
```diff
@@ -290,9 +290,372 @@ def _on_expired_window(
...
+        timestamp_ms: int,
+        late_by_ms: int,
+    ) -> None:
+        try:
+            ctx = message_context()
+            topic = ctx.topic
+            partition = ctx.partition
+            offset = ctx.offset
+        except:
+            # In test environments, message context might not be available
```
Comment (gwaramadze) [inline quixstreams/dataframe/windows/time_based.py:615]: We must not mock objects for test purposes conditionally in the lib code. Sliding windows have `mock_message_context`, so it should be reused where necessary.

File: quixstreams/dataframe/windows/time_based.py [hunk 5]
```diff
@@ -290,9 +290,372 @@ def _on_expired_window(
...
+                state.delete_from_collection(window_end, start=window_start)
+
+        if count:
+            logger.debug(
+                "Expired %s session windows in %ss",
+                count,
+                round(time.monotonic() - start, 2),
+            )
+
+    def _on_expired_session(
```
Comment (gwaramadze) [inline quixstreams/dataframe/windows/time_based.py:600]: This method is almost identical to existing `_on_expired_window`, we should abstract it.

File: docs/windowing.md [hunk 6]
```diff
@@ -500,6 +502,220 @@ sdf = (
...
+    )
+
+    # Emit results when sessions are complete
+    .final()
+)
+
+# Expected output (when session expires):
+# {
+#   "start": 1000,
+#   "end": 2000000 + 1800000,  # last event + timeout
```
Comment (gwaramadze) [inline docs/windowing.md:581]: The start of the window is rightfully assigned to the timestamp of the first message in a session. Analogously, the end of the window should be assigned to the timestamp of the last message in the session, rather than being extended by the full length of a timeout.

File: quixstreams/dataframe/windows/time_based.py [hunk 7]
```diff
@@ -290,9 +290,372 @@ def _on_expired_window(
...
+                id=timestamp_ms,
+            )
+
+        # Update the session window aggregation
+        aggregated = None
+        if aggregate:
+            current_value = existing_aggregated if can_extend_session else None
+            if current_value is None:
+                current_value = self._initialize_value()
+
```
Comment (gwaramadze) [inline quixstreams/dataframe/windows/time_based.py:476]:
```suggestion
            current_value = (
                existing_aggregated if can_extend_session else self._initialize_value()
            )
```

File: quixstreams/dataframe/windows/time_based.py [hunk 8]
```diff
@@ -290,9 +290,372 @@ def _on_expired_window(
...
+            aggregated = self._aggregate_value(current_value, value, timestamp_ms)
+
+        # By this point, session_start and session_end are guaranteed to be set
+        assert session_start is not None  # noqa: S101
+        assert session_end is not None  # noqa: S101
+
+        # Output intermediate results for aggregations
+        if aggregate:
```
Comment (gwaramadze) [inline quixstreams/dataframe/windows/time_based.py:484]: Asserts are unnecessary. If we let them go, the next `if` is also rendered redundant.
```suggestion
```

File: quixstreams/dataframe/windows/time_based.py [hunk 9]
```diff
@@ -290,9 +290,372 @@ def _on_expired_window(
...
+                windows_to_delete.append((window_start, window_end))
+                count += 1
+                yield (
+                    key,
+                    self._results(aggregated, collected, window_start, window_end),
+                )
+
+        # Clean up expired windows
+        for window_start, window_end in windows_to_delete:
+            state._transaction.delete_window(  # type: ignore # noqa: SLF001
```
Comment (gwaramadze) [inline quixstreams/dataframe/windows/time_based.py:585]: Usage of the private attribute `_transaction`, this must be fixed.

File: quixstreams/dataframe/dataframe.py [hunk 10]
```diff
@@ -1484,6 +1485,103 @@ def sliding_count_window(
             name=name,
         )
 
+    def session_window(
+        self,
+        timeout_ms: Union[int, timedelta],
```
Comment (gwaramadze) [inline quixstreams/dataframe/dataframe.py:1490]: A more widely adopted nomenclature for session windows is "gap". Let's rename it to `gap_ms` or a more explicit `inactivity_gap_ms`.

File: quixstreams/dataframe/windows/aggregations.py [hunk 11]
```diff
@@ -348,6 +348,7 @@ def __init__(
         reducer: Callable[[R, Any], R],
         initializer: Callable[[Any], R],
     ) -> None:
+        super().__init__()
```
Comment (gwaramadze) [inline quixstreams/dataframe/windows/aggregations.py:351]: Wasn't this a part of a separate bugfix PR? If not, then it should be separated together with the corresponding test case.

File: quixstreams/dataframe/windows/time_based.py [hunk 12]
```diff
@@ -290,9 +290,372 @@ def _on_expired_window(
...
+                window_start,
+                window_end,
+                prefix=state._prefix,  # type: ignore # noqa: SLF001
```
Comment (gwaramadze) [inline quixstreams/dataframe/windows/time_based.py:588]: also here, we can't use the private member like this

**Review by gwaramadze — COMMENTED** [review]
Hello. Once again, thanks for your contributions. Here is the first batch of comments. Please take a look. We will continue to review and test it.

**Comment by gwaramadze** [comment]
Closing this PR as it was superseded by https://github.com/quixio/quix-streams/pull/994
